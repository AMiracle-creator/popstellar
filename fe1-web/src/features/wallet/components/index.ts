import RollCallTokensDropDown from './RollCallTokensDropDown';
import RoundIconButton from './RoundIconButton';

export { RollCallTokensDropDown, RoundIconButton };
