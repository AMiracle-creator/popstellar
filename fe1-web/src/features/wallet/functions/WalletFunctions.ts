import { WalletStore } from '../store';

export const { hasSeed } = WalletStore;

export { generateToken } from '../objects';
