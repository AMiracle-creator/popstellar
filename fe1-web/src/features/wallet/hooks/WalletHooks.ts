import { useContext } from 'react';

import FeatureContext from 'core/contexts/FeatureContext';

import { WalletReactContext, WALLET_FEATURE_IDENTIFIER } from '../interface';

export namespace WalletHooks {
  export const useWalletContext = (): WalletReactContext => {
    const featureContext = useContext(FeatureContext);
    // assert that the wallet context exists
    if (!(WALLET_FEATURE_IDENTIFIER in featureContext)) {
      throw new Error('Wallet context could not be found!');
    }
    return featureContext[WALLET_FEATURE_IDENTIFIER] as WalletReactContext;
  };

  /**
   * Gets the list of wallet item generators
   */
  export const useWalletItemGenerators = () => useWalletContext().walletItemGenerators;

  /**
   * Gets the current lao id
   */
  export const useCurrentLaoId = () => useWalletContext().useCurrentLaoId();

  /**
   * Returns true if currently connected to a lao, false if in offline mode
   * and undefined if there is no current lao
   */
  export const useConnectedToLao = () => useWalletContext().useConnectedToLao();

  /**
   * Gets all lao ids
   */
  export const useLaoIds = () => useWalletContext().useLaoIds();

  /**
   * Gets a map from rollCall ids to rollCall instances for a given lao id
   */
  export const useRollCallsByLaoId = (laoId: string) =>
    useWalletContext().useRollCallsByLaoId(laoId);

  /**
   * Gets the function for obtain roll call tokens by lao id
   */
  export const useRollCallTokensByLaoId = (laoId: string) =>
    useWalletContext().useRollCallTokensByLaoId(laoId);

  /**
   * Gets a map from laoIds to names
   */
  export const useNamesByLaoId = () => useWalletContext().useNamesByLaoId();

  /**
   * Gets the list of wallet navigation screens
   */
  export const useWalletNavigationScreens = () => useWalletContext().walletNavigationScreens;
}
