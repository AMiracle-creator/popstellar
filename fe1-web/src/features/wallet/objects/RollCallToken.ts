import { Hash, KeyPairState, PopToken } from 'core/objects';

export interface RollCallTokenState {
  token: KeyPairState;
  laoId: string;
  rollCallId: string;
  rollCallName: string;
}
/**
 * A Roll Call Token object, defined by a Pop token, its lao id hash and its Roll Call id hash
 */
export class RollCallToken {
  public readonly token: PopToken;

  public readonly laoId: Hash;

  public readonly rollCallId: Hash;

  public readonly rollCallName: string;

  constructor(obj: Partial<RollCallToken>) {
    if (obj === undefined || obj === null) {
      throw new Error(
        'Error encountered while creating a RollCallToken object: undefined/null parameters',
      );
    }

    if (obj.token === undefined) {
      throw new Error("Undefined 'token' when creating 'RollCallToken'");
    }
    if (obj.laoId === undefined) {
      throw new Error("Undefined 'laoId' when creating 'RollCallToken'");
    }
    if (obj.rollCallId === undefined) {
      throw new Error("Undefined 'rollCallId' when creating 'RollCallToken'");
    }
    if (obj.rollCallName === undefined) {
      throw new Error("Undefined 'rollCallName' when creating 'RollCallToken'");
    }

    this.token = obj.token;
    this.laoId = obj.laoId;
    this.rollCallId = obj.rollCallId;
    this.rollCallName = obj.rollCallName;
  }

  public static fromState(rct: RollCallTokenState): RollCallToken {
    return new RollCallToken({
      token: PopToken.fromState(rct.token),
      laoId: new Hash(rct.laoId),
      rollCallId: new Hash(rct.rollCallId),
      rollCallName: rct.rollCallName,
    });
  }

  public toState(): RollCallTokenState {
    return {
      token: this.token.toState(),
      laoId: this.laoId.valueOf(),
      rollCallId: this.rollCallId.valueOf(),
      rollCallName: this.rollCallName,
    };
  }
}
