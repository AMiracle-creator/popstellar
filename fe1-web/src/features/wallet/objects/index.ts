export * from './Seed';
export { generateToken, getCurrentPopTokenFromStore } from './Token';
