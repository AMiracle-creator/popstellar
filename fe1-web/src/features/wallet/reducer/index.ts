export { default as walletReducer } from './WalletReducer';
export { default as digitalCashReducer } from '../../digital-cash/reducer/DigitalCashReducer';
export * from './WalletReducer';
export * from '../../digital-cash/reducer/DigitalCashReducer';
