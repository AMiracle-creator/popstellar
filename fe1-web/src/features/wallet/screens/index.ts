export { default as WalletHome } from './WalletHome';
export { default as WalletSetSeed } from './WalletSetSeed';
export { default as WalletCreateSeed } from './WalletCreateSeed';
