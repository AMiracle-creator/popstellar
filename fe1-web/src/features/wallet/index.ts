import { SignatureType } from 'core/network/jsonrpc/messages';

import * as functions from './functions';
import {
  WalletCompositionConfiguration,
  WalletCompositionInterface,
  WalletInterface,
  WALLET_FEATURE_IDENTIFIER,
} from './interface';
import * as navigation from './navigation';
import { configureNetwork } from './network';
import { getCurrentPopTokenFromStore } from './objects';
import { digitalCashReducer, walletReducer } from './reducer';

/**
 * Configures the wallet feature
 */
export function configure(): WalletInterface {
  return {
    identifier: WALLET_FEATURE_IDENTIFIER,
    functions,
  };
}

export function compose(configuration: WalletCompositionConfiguration): WalletCompositionInterface {
  configureNetwork(configuration.messageRegistry);
  configuration.keyPairRegistry.add(
    SignatureType.POP_TOKEN,
    getCurrentPopTokenFromStore(configuration.getCurrentLao, configuration.getRollCallById),
  );
  return {
    identifier: WALLET_FEATURE_IDENTIFIER,
    navigation,
    reducers: {
      ...walletReducer,
      ...digitalCashReducer,
    },
    context: {
      useRollCallsByLaoId: configuration.useRollCallsByLaoId,
      useCurrentLaoId: configuration.useCurrentLaoId,
      getLaoOrganizer: configuration.getLaoOrganizer,
    },
  };
}
