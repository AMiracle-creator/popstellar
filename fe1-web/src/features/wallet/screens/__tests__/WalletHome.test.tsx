import { render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers, createStore } from 'redux';

import MockNavigator from '__tests__/components/MockNavigator';
import { mockKeyPair, mockLao, mockLaoId, mockLaoIdHash, mockPopToken } from '__tests__/utils';
import FeatureContext from 'core/contexts/FeatureContext';
import { EventTags, Hash, Timestamp } from 'core/objects';
import { getEventById } from 'features/events/functions';
import { addEvent, eventReducer, makeEventByTypeSelector } from 'features/events/reducer';
import { RollCallHooks } from 'features/rollCall/hooks';
import { RollCallReactContext, ROLLCALL_FEATURE_IDENTIFIER } from 'features/rollCall/interface';
import { RollCall, RollCallStatus } from 'features/rollCall/objects';
import { addRollCall, rollCallReducer } from 'features/rollCall/reducer';
import { hasSeed } from 'features/wallet/functions';
import { WalletReactContext, WALLET_FEATURE_IDENTIFIER } from 'features/wallet/interface';
import { walletReducer } from 'features/wallet/reducer';
import STRINGS from 'resources/strings';

import { generateToken, recoverWalletRollCallTokens } from '../../objects';
import { RollCallToken } from '../../objects/RollCallToken';
import { WalletHome } from '../index';

jest.mock('core/platform/Storage');
jest.mock('core/platform/crypto/browser');
jest.mock('features/wallet/objects/Wallet');
jest.mock('core/components/QRCode.tsx', () => 'qrcode');

const mockRCName = 'myRollCall';
const mockRCLocation = 'location';
const mockRCTimestampStart = new Timestamp(1620255600);
const mockRCTimestampEnd = new Timestamp(1620357600);
const mockRCAttendees = ['attendee1', 'attendee2'];

const mockRCIdAliasHash = Hash.fromStringArray(
  EventTags.ROLL_CALL,
  mockLaoId,
  mockRCTimestampStart.toString(),
  mockRCName,
);

const mockRCIdHash = Hash.fromStringArray(
  EventTags.ROLL_CALL,
  mockLaoId,
  mockRCIdAliasHash.valueOf(),
  mockRCName,
);

const mockRollCallState = {
  id: mockRCIdHash.valueOf(),
  idAlias: mockRCIdAliasHash.valueOf(),
  eventType: RollCall.EVENT_TYPE,
  start: mockRCTimestampStart.valueOf(),
  end: mockRCTimestampEnd.valueOf(),
  name: mockRCName,
  location: mockRCLocation,
  creation: mockRCTimestampStart.valueOf(),
  proposedStart: mockRCTimestampStart.valueOf(),
  proposedEnd: mockRCTimestampEnd.valueOf(),
  status: RollCallStatus.CLOSED,
  attendees: mockRCAttendees,
};
const mockRollCall = RollCall.fromState(mockRollCallState);

const contextValue = {
  [WALLET_FEATURE_IDENTIFIER]: {
    useCurrentLaoId: () => mockLaoIdHash,
    getEventById,
    useRollCallsByLaoId: RollCallHooks.useRollCallsByLaoId,
    getLaoOrganizer: jest.fn(() => mockKeyPair.publicKey),
  } as WalletReactContext,
  [ROLLCALL_FEATURE_IDENTIFIER]: {
    useCurrentLaoId: () => mockLaoIdHash,
    generateToken,
    hasSeed,
    makeEventByTypeSelector,
  } as RollCallReactContext,
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe('Wallet home', () => {
  it('renders correctly with an empty wallet', () => {
    const mockStore = createStore(
      combineReducers({ ...walletReducer, ...rollCallReducer, ...eventReducer }),
    );

    const component = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={WalletHome} />
        </FeatureContext.Provider>
      </Provider>,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with a non empty wallet', async () => {
    const mockStore = createStore(
      combineReducers({ ...walletReducer, ...rollCallReducer, ...eventReducer }),
    );

    const mockRCToken = new RollCallToken({
      token: mockPopToken,
      laoId: mockLao.id,
      rollCallId: mockRollCall.id,
      rollCallName: mockRollCall.name,
    });

    // make the selector return data
    mockStore.dispatch(
      addEvent(mockLaoId, {
        eventType: RollCall.EVENT_TYPE,
        id: mockRollCallState.id,
        start: mockRollCallState.proposedStart,
        end: mockRollCallState.proposedEnd,
      }),
    );
    mockStore.dispatch(addRollCall(mockRollCallState));

    (recoverWalletRollCallTokens as jest.Mock).mockImplementation(() =>
      Promise.resolve([mockRCToken]),
    );

    const component = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={WalletHome} />
        </FeatureContext.Provider>
      </Provider>,
    );

    expect(recoverWalletRollCallTokens).toHaveBeenCalledTimes(1);
    expect(recoverWalletRollCallTokens).toHaveBeenCalledWith(expect.anything(), mockLaoIdHash);

    await waitFor(() => {
      expect(() => component.getByText(STRINGS.no_tokens_in_wallet)).toThrow();
    });

    expect(component).toMatchSnapshot();
  });
});
