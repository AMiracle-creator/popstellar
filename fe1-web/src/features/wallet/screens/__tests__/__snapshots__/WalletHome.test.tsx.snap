// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet home renders correctly with a non empty wallet 1`] = `null`;

exports[`Wallet home renders correctly with an empty wallet 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              nativeID="animatedComponent"
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            nativeID="animatedComponent"
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  nativeID="animatedComponent"
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  MockScreen
                </Text>
              </View>
              <View
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              nativeID="animatedComponent"
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                            "padding": 30,
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "display": "flex",
                              "flexDirection": "row",
                              "justifyContent": "space-between",
                              "paddingLeft": 20,
                              "paddingRight": 20,
                              "width": "100%",
                            }
                          }
                        >
                          <Text
                            style={
                              Object {
                                "fontSize": 25,
                                "fontWeight": "bold",
                                "textAlign": "left",
                              }
                            }
                          >
                            Welcome to your Digital Wallet
                          </Text>
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            nativeID="animatedComponent"
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "alignContent": "center",
                                "alignItems": "center",
                                "backgroundColor": "#2196F3",
                                "borderRadius": 30,
                                "height": 65,
                                "justifyContent": "center",
                                "opacity": 1,
                                "shadowColor": "gray",
                                "shadowOffset": Object {
                                  "height": 2,
                                  "width": 2,
                                },
                                "width": 65,
                              }
                            }
                          >
                            <Text />
                          </View>
                        </View>
                        <View
                          style={
                            Object {
                              "flexDirection": "row",
                              "marginBottom": 20,
                              "marginTop": 30,
                            }
                          }
                        />
                        <Text
                          style={
                            Object {
                              "fontSize": 25,
                              "marginHorizontal": 10,
                              "textAlign": "center",
                            }
                          }
                        >
                          No token is yet associated with your wallet seed, once you participate in a roll call event your PoP tokens will be shown here
                        </Text>
                        <Modal
                          animationType="slide"
                          hardwareAccelerated={false}
                          onRequestClose={[Function]}
                          visible={false}
                        >
                          <View
                            style={
                              Object {
                                "height": "80%",
                                "padding": 20,
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "alignItems": "center",
                                  "flex": 1,
                                  "marginHorizontal": 50,
                                  "marginVertical": 10,
                                }
                              }
                            >
                              <View
                                style={
                                  Object {
                                    "flexDirection": "row",
                                    "justifyContent": "space-between",
                                    "width": "100%",
                                  }
                                }
                              >
                                <Text
                                  style={
                                    Object {
                                      "fontSize": 25,
                                      "fontWeight": "bold",
                                      "textAlign": "left",
                                    }
                                  }
                                >
                                  Send
                                </Text>
                                <View
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  nativeID="animatedComponent"
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                  style={
                                    Object {
                                      "alignItems": "center",
                                      "backgroundColor": "#2196F3",
                                      "borderRadius": 30,
                                      "display": "flex",
                                      "height": 80,
                                      "justifyContent": "center",
                                      "opacity": 1,
                                      "shadowColor": "gray",
                                      "shadowOffset": Object {
                                        "height": 2,
                                        "width": 2,
                                      },
                                      "width": 80,
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "overflow": "hidden",
                                      }
                                    }
                                  >
                                    <View>
                                      <View
                                        style={
                                          Object {
                                            "alignItems": "center",
                                            "backgroundColor": "transparent",
                                            "justifyContent": "center",
                                          }
                                        }
                                      >
                                        <Text />
                                      </View>
                                    </View>
                                  </View>
                                </View>
                              </View>
                              <View
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "display": "flex",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "maxWidth": "100%",
                                  }
                                }
                              >
                                <View
                                  style={
                                    Object {
                                      "margin": "auto",
                                      "maxWidth": "100%",
                                    }
                                  }
                                >
                                  <View
                                    style={
                                      Object {
                                        "paddingHorizontal": 10,
                                        "width": "100%",
                                      }
                                    }
                                  >
                                    <Text
                                      style={
                                        Object {
                                          "color": "#86939e",
                                          "fontSize": 16,
                                          "fontWeight": "bold",
                                        }
                                      }
                                    >
                                      Receiver's public key
                                    </Text>
                                    <View
                                      collapsable={false}
                                      nativeID="animatedComponent"
                                      style={
                                        Object {
                                          "alignItems": "center",
                                          "borderBottomWidth": 1,
                                          "borderColor": "#86939e",
                                          "flexDirection": "row",
                                          "transform": Array [
                                            Object {
                                              "translateX": 0,
                                            },
                                          ],
                                        }
                                      }
                                    >
                                      <TextInput
                                        editable={true}
                                        onChangeText={[Function]}
                                        placeholderTextColor="#86939e"
                                        style={
                                          Object {
                                            "color": "#242424",
                                            "flex": 1,
                                            "fontSize": 18,
                                            "minHeight": 40,
                                            "minWidth": 400,
                                            "padding": 5,
                                          }
                                        }
                                        testID="RNE__Input__text-input"
                                        underlineColorAndroid="transparent"
                                      />
                                    </View>
                                    <Text
                                      style={
                                        Object {
                                          "color": "#ff190c",
                                          "fontSize": 12,
                                          "margin": 5,
                                        }
                                      }
                                    />
                                  </View>
                                  <View
                                    style={
                                      Object {
                                        "paddingHorizontal": 10,
                                        "width": "100%",
                                      }
                                    }
                                  >
                                    <Text
                                      style={
                                        Object {
                                          "color": "#86939e",
                                          "fontSize": 16,
                                          "fontWeight": "bold",
                                        }
                                      }
                                    >
                                      Amount to send
                                    </Text>
                                    <View
                                      collapsable={false}
                                      nativeID="animatedComponent"
                                      style={
                                        Object {
                                          "alignItems": "center",
                                          "borderBottomWidth": 1,
                                          "borderColor": "#86939e",
                                          "flexDirection": "row",
                                          "transform": Array [
                                            Object {
                                              "translateX": 0,
                                            },
                                          ],
                                        }
                                      }
                                    >
                                      <TextInput
                                        editable={true}
                                        onChangeText={[Function]}
                                        placeholderTextColor="#86939e"
                                        style={
                                          Object {
                                            "color": "#242424",
                                            "flex": 1,
                                            "fontSize": 18,
                                            "minHeight": 40,
                                            "minWidth": 400,
                                            "padding": 5,
                                          }
                                        }
                                        testID="RNE__Input__text-input"
                                        underlineColorAndroid="transparent"
                                      />
                                    </View>
                                    <Text
                                      style={
                                        Object {
                                          "color": "#ff190c",
                                          "fontSize": 12,
                                          "margin": 5,
                                        }
                                      }
                                    />
                                  </View>
                                  <View
                                    style={
                                      Object {
                                        "flexDirection": "row",
                                        "margin": 15,
                                      }
                                    }
                                  >
                                    <Text>
                                      Coinbase: 
                                    </Text>
                                    <RCTSwitch
                                      accessibilityRole="switch"
                                      onChange={[Function]}
                                      onResponderTerminationRequest={[Function]}
                                      onStartShouldSetResponder={[Function]}
                                      onTintColor="#81b0ff"
                                      style={
                                        Array [
                                          Object {
                                            "height": 31,
                                            "width": 51,
                                          },
                                          Object {
                                            "backgroundColor": "#3e3e3e",
                                            "borderRadius": 16,
                                          },
                                        ]
                                      }
                                      thumbTintColor="#f4f3f4"
                                      tintColor="#767577"
                                      value={false}
                                    />
                                  </View>
                                  <View
                                    accessibilityRole="button"
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    nativeID="animatedComponent"
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                    style={
                                      Object {
                                        "opacity": 1,
                                      }
                                    }
                                  >
                                    <View
                                      style={
                                        Array [
                                          Object {},
                                        ]
                                      }
                                    >
                                      <Text
                                        style={
                                          Array [
                                            Object {
                                              "color": "#007AFF",
                                              "fontSize": 18,
                                              "margin": 8,
                                              "textAlign": "center",
                                            },
                                          ]
                                        }
                                      >
                                        Send
                                      </Text>
                                    </View>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </Modal>
                        <View
                          style={
                            Object {
                              "padding": "1rem",
                            }
                          }
                        />
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
